sdk/dotnet/Pulumi.Automation/Events/EventLogWatcher.cs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2016-2021, Pulumi Corporation
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Pulumi.Automation.Serialization;
+
+namespace Pulumi.Automation.Events
+{
+    internal class EventLogWatcher : IAsyncDisposable
+    {
+        private readonly LocalSerializer _localSerializer = new LocalSerializer();
+        private readonly CancellationTokenSource _internalCancellationTokenSource = new CancellationTokenSource();
+        private readonly CancellationToken _externalCancellationToken;
+        private readonly Action<EngineEvent> _onEvent;
+        private readonly Task _pollingTask;
+
+        private const int _pollingIntervalMilliseconds = 100;
+
+        // We keep track of the length we have already read from the file
+        private long _previousLength = 0;
+
+        public string LogFile { get; }
+
+        internal EventLogWatcher(
+            string logFile
+            , Action<EngineEvent> onEvent
+            , CancellationToken externalCancellationToken)
+        {
+            LogFile = logFile;
+            _onEvent = onEvent;
+            _externalCancellationToken = externalCancellationToken;
+            _pollingTask = Task.Run(PollForEvents);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            _internalCancellationTokenSource.Cancel();
+            await _pollingTask.ConfigureAwait(false);
+        }
+
+        private async Task PollForEvents()
+        {
+            var internalToken = _internalCancellationTokenSource.Token;
+            var externalToken = _externalCancellationToken;
+
+            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(internalToken, externalToken);
```

Comment:
can you help me understand the function of the internal CTS?

Reply:
The internal CTS allows us to stop polling when the watcher is being disposed by calling its `Cancel` method. Using `CreateLinkedTokenSource` ensures that we respect both our internal CTS as well as the user provided token, so if the 
user token ends up requesting cancellation we also stop polling.

Reply:
hmmm OK well that makes sense functionally. I wonder if it could just be another boolean. The internal CTS needs to be disposed as well

Reply:
A boolean flag could work, but it would mean that we'd end up waiting for the `Task.Delay` to finish before stopping, with cancelling the CTS the Task.Delay will throw `TaskCanceledException` which we catch to `break` from the polling. If the polling interval is sufficiently small, I don't think this makes much of a practical difference, but if the polling interval was sufficiently large, for example a second or more we would end up waiting longer before stopping.

Reply:
no you're right this is fine - For some reason I was thinking that disposing the internal CTS so soon after cancelling would cause issues but since you are awaiting the polling task before disposing it is perfectly fine

Reply:
I've enjoyed the CreateLinkedTokenSource, seems like a great find that fits the problem well.